File: mobile/ios/packages/town-crier-presentation/Sources/Coordinators/ReviewPromptTracker.swift
import Foundation
import TownCrierDomain

/// Performs the real App Store review request. Abstracted behind a protocol so
/// the decision flow stays unit-testable (the OS call sits at the app-layer
/// edge — see ``CoordinatorReviewRequester``).
@MainActor
public protocol ReviewRequesting {
  func requestReview()
}

/// Production ``ReviewRequesting`` that flips ``AppCoordinator/isRequestingReview``.
///
/// An app-layer `.onChange(of:)` observes that flag, performs the real
/// `requestReview` via the SwiftUI environment, then resets it — mirroring the
/// established `isOpeningSystemNotificationSettings` → `openURL` pattern. The
/// coordinator reference is weak to avoid a retain cycle (the coordinator owns
/// the tracker, which owns this requester).
@MainActor
public final class CoordinatorReviewRequester: ReviewRequesting {
  public weak var coordinator: AppCoordinator?

  public init(coordinator: AppCoordinator? = nil) {
    self.coordinator = coordinator
  }

  public func requestReview() {
    coordinator?.isRequestingReview = true
  }
}

/// The @MainActor service the app talks to for the App Store review prompt
/// (GH #628).
///
/// It owns the only mutable session flag (``suppressThisSession()``), reads and
/// writes the persisted state via the injected ``ReviewPromptStore``, runs the
/// pure ``ReviewPromptPolicy`` on each signal, and — on a `.fire` decision —
/// asks the injected ``ReviewRequesting`` to present the dialog. Being
/// `@MainActor`, back-to-back signals are serialized, so each is evaluated once.
@MainActor
public final class ReviewPromptTracker {
  private let store: ReviewPromptStore
  private let policy: ReviewPromptPolicy
  private let requester: any ReviewRequesting
  private let now: @Sendable () -> Date
  private var isSessionSuppressed = false

  public init(
    store: ReviewPromptStore,
    requester: any ReviewRequesting,
    now: @escaping @Sendable () -> Date = Date.init,
    policy: ReviewPromptPolicy? = nil
  ) {
    self.store = store
    self.requester = requester
    self.now = now
    self.policy = policy ?? ReviewPromptPolicy(now: now)
    establishFirstLaunchDateIfNeeded()
  }

  /// Suppresses any review prompt for the rest of this session. Called during
  /// onboarding, when the post-purchase push prompt fires, and on friction.
  public func suppressThisSession() {
    isSessionSuppressed = true
  }

  /// Records an engagement signal: updates the persisted state via the policy
  /// and, on a `.fire`, requests the native review dialog.
  public func record(_ signal: ReviewSignal, isReactivation: Bool = false) {
    let outcome = policy.evaluate(
      signal: signal,
      state: store.load(),
      sessionSuppressed: isSessionSuppressed,
      isReactivation: isReactivation
    )
    store.save(outcome.state)
    if outcome.decision == .fire {
      requester.requestReview()
    }
  }

  /// Records a loyalty active-day signal on app foreground. `isReactivation` is
  /// `true` only for a background→active re-entry (never the cold-launch render).
  public func recordAppForegrounded(isReactivation: Bool) {
    record(.activeDay, isReactivation: isReactivation)
  }

  /// Anchors the account-age guard the first time the tracker runs on a device.
  private func establishFirstLaunchDateIfNeeded() {
    var state = store.load()
    guard state.firstLaunchDate == nil else { return }
    state.firstLaunchDate = now()
    store.save(state)
  }
}

File: mobile/android/presentation/src/main/kotlin/uk/towncrierapp/presentation/reviewprompt/ReviewPromptTracker.kt
package uk.towncrierapp.presentation.reviewprompt

import uk.towncrierapp.domain.reviewprompt.ReviewPromptDecision
import uk.towncrierapp.domain.reviewprompt.ReviewPromptPolicy
import uk.towncrierapp.domain.reviewprompt.ReviewPromptState
import uk.towncrierapp.domain.reviewprompt.ReviewPromptStore
import uk.towncrierapp.domain.reviewprompt.ReviewRequester
import uk.towncrierapp.domain.reviewprompt.ReviewSignal
import java.time.Clock

/**
 * The service the app talks to for the store review prompt (GH #628). Owns
 * the only mutable session flag ([suppressThisSession]), reads and writes
 * the persisted state via the injected [ReviewPromptStore], runs the pure
 * [ReviewPromptPolicy] on each signal, and — on a FIRE decision — asks the
 * injected [ReviewRequester] to present the dialog. Port of iOS
 * `ReviewPromptTracker`.
 */
public class ReviewPromptTracker(
    private val store: ReviewPromptStore,
    private val requester: ReviewRequester,
    private val clock: Clock = Clock.systemUTC(),
    private val policy: ReviewPromptPolicy = ReviewPromptPolicy(clock),
) {
    private var isSessionSuppressed = false

    /**
     * Suppresses any review prompt for the rest of this session. Call during
     * onboarding, when a post-purchase prompt fires, and on friction.
     */
    public fun suppressThisSession() {
        isSessionSuppressed = true
    }

    /**
     * Records an engagement signal: updates the persisted state via the
     * policy and, on a FIRE decision, requests the native review dialog.
     */
    public suspend fun recordSignal(
        signal: ReviewSignal,
        isReactivation: Boolean = false,
    ) {
        val state = establishFirstLaunchDateIfNeeded(store.load())
        val outcome = policy.evaluate(signal, state, isSessionSuppressed, isReactivation)
        store.save(outcome.state)
        if (outcome.decision == ReviewPromptDecision.FIRE) {
            requester.requestReview()
        }
    }

    /**
     * Records a loyalty active-day signal on app foreground. [isReactivation]
     * is `true` only for a background-to-active re-entry, never the
     * cold-launch render.
     */
    public suspend fun recordAppForegrounded(isReactivation: Boolean) {
        recordSignal(ReviewSignal.ActiveDay, isReactivation)
    }

    /** Anchors the account-age guard the first time the tracker runs on a device. */
    private suspend fun establishFirstLaunchDateIfNeeded(state: ReviewPromptState): ReviewPromptState {
        if (state.firstLaunchDate != null) return state
        val updated = state.copy(firstLaunchDate = clock.instant())
        store.save(updated)
        return updated
    }
}
